activerecord/test/cases/base_test.rb

```diff
@@ -1012,7 +1012,7 @@ def test_dup_for_a_composite_primary_key_model
 
     assert_equal "The first book", new_book.title
     assert_nil new_book.author_id
-    assert_nil new_book.number
+    assert_nil new_book._read_attribute("id")
```

Comment:
Nitpick but sometimes tests may be treated as API examples and ideally we don't want `_read_attribute` to be used an example. Perhaps we could do something like `assert_equal [nil, nil], new_book.id` since we want to assert both parts of the PK

Reply:
Can you change those tests to assert the `id` though?

Reply:
I mean, if documentation wasn't sufficient and I were to get an least some brief understanding of how logic is supposed to be used I'd look into tests. My main motivation is that often inability to test something using public interfaces means that public interface might not be complete or perhaps we are testing something private. In most cases it happens in tests that are forced to use `instance_variable_get `

For example a test like this may not be a formal API
https://github.com/rails/rails/blob/fe52de644b1c3a3d60d160c52b5a31b28cb42645/activerecord/test/cases/attribute_methods_test.rb#L457
but provides a lot of insight into what is the current expectation, current spec of predicate behaviors. 

So yea, not an API examples but still something that can be expected to be used as an insight so it would be better aim to use best practises